### PR TITLE
Migrate codebase to Typescript

### PR DESCRIPTION
**Pull request by AjayPoshak**
## Summary
Moved codebase to Typescript from Flow JS.  TS is fun so far 💯 

## Related Issue #[issue number]
Related Issue #111 

## Any Breaking Changes
Nope, It won't impact end user

## Checklist
- [X] Are all the test cases passing?
- [X] If any new feature has been added, then are the test cases updated/added?
- [X] Has the documentation been updated for the proposed change, if required?

### CODE REVIEW

File: src/Svg.tsx [hunk 1]
```diff
@@ -1,38 +1,39 @@
...
-import uid from "./uid"
-import type { Props as HolderProps } from "./Holder"
+import { IProps } from "./Holder";
 
-export type SvgProps = {
-  children?: React.ChildrenArray<*>
-} & HolderProps
+import uid from "./uid";
+
+export interface ISvgProps extends IProps {
```
Comment (danilowoz) [inline src/Svg.tsx]: Here all props should be required, so I don't know if there is a way to "convert" all types in required or we should create a new Interface with all props again, but required. Do you have any suggestion for that?

File: src/index.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import Holder from "./Holder";
+
```
[the remark this answers is not in the document]
Comment (danilowoz) [reply src/index.ts]: Great

File: src/stylized/BulletListStyle.tsx [hunk 3]
```diff
@@ -1,9 +1,8 @@
-// @flow
-import * as React from "react"
-import Holder from "../Holder"
-import type { SvgProps } from "../Svg"
+import * as React from "react";
+import Holder from "../Holder";
+import { ISvgProps } from "../Svg";
 
-const BulletListStyle = (props: SvgProps): React.Element<*> => (
+const BulletListStyle = (props: ISvgProps) => (
```
Comment (danilowoz) [inline src/stylized/BulletListStyle.tsx]: At same thing here, now all props should be required like `Svg`

File: src/stylized/BulletListStyle.tsx [hunk 4]
```diff
@@ -1,9 +1,8 @@
-// @flow
-import * as React from "react"
-import Holder from "../Holder"
-import type { SvgProps } from "../Svg"
+import * as React from "react";
```
Comment (danilowoz) [inline src/stylized/BulletListStyle.tsx]: Break the line after `react` import

File: src/stylized/BulletListStyle.tsx [hunk 5]
```diff
@@ -14,6 +13,6 @@ const BulletListStyle = (props: SvgProps): React.Element<*> => (
     <circle cx="10" cy="110" r="8" />
     <rect x="25" y="105" rx="5" ry="5" width="220" height="10" />
   </Holder>
-)
+);
 
-export default BulletListStyle
+export default BulletListStyle;
```
Comment (danilowoz) [inline src/stylized/BulletListStyle.tsx]: I think it's a great time to remove the export from here and put in the definition of component, like:

```tsx
export default (props: ISvgProps) => (
```

Comment (danilowoz) [inline src/stylized/BulletListStyle.tsx]: Please apply these changes in the others components

File: tslint.json [hunk 6]
```diff
@@ -0,0 +1,9 @@
+{
+  "defaultSeverity": "error",
+  "extends": [
+      "tslint:recommended"
```
Comment (danilowoz) [inline tslint.json]: Add the `tslint-config-prettier` to work fine with prettier

File: package.json [hunk 7]
```diff
@@ -33,26 +34,28 @@
...
-    "test:watch": "npm run testonly -- --watch --watch-extensions js",
+    "test": "npm run lint && npm run testonly && npm run tsc",
+    "test:watch": "npm run testonly -- --watch --watch-extensions ts",
     "testonly": "cross-env NODE_ENV=test mocha $npm_package_options_mocha",
     "release:major": "npm run build && npm version major",
     "release:minor": "npm run build && npm version minor",
     "release:patch": "npm run build && npm version patch",
     "lint": "eslint 'src/**/*.js'",
-    "flow": "flow",
+    "watch": "node_modules/.bin/tsc --watch",
```
Comment (danilowoz) [inline package.json]: Missing `tslint` dependency and script. 
Once added the `tslint` you can remove `eslint` stuff

File: package.json [hunk 8]
```diff
@@ -72,12 +74,13 @@
     "react": "^16.1.1",
     "react-dom": "^16.1.1",
     "rollup": "0.66.4",
-    "rollup-plugin-babel": "^4.0.3",
     "rollup-plugin-replace": "^2.0.0",
-    "rollup-plugin-uglify": "^6.0.0"
+    "rollup-plugin-typescript2": "^0.18.0",
+    "rollup-plugin-uglify": "^6.0.0",
+    "ts-node": "^7.0.1",
```
Comment (danilowoz) [inline package.json]: why exactly we need this?

File: src/Svg.tsx [hunk 9]
```diff
@@ -1,38 +1,39 @@
...
   primaryOpacity,
-  rtl,
   secondaryColor,
   secondaryOpacity,
-  speed,
-  style,
-  uniquekey,
-  width,
+  preserveAspectRatio,
+  // tslint:disable-next-line:trailing-comma
```
Comment (danilowoz) [inline src/Svg.tsx]: Is it really necessary? I think with the suggestion in the `tslint` that I did, it'll be solved

File: src/Holder.tsx [hunk 10]
```diff
@@ -0,0 +1,49 @@
+import * as React from "react";
+import Svg from "./Svg";
+
+export interface IProps {
```
[the remark this answers is not in the document]
Comment (ifiokjr) [reply src/Holder.tsx]: Sure thing 👍

File: src/Holder.tsx [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+  primaryColor: "#f0f0f0",
+  primaryOpacity: 1,
+  rtl: false,
+  secondaryColor: "#e0e0e0",
+  secondaryOpacity: 1,
+  speed: 2,
+  width: 400,
+};
+
+const InitialComponent = (props: IProps) => (
```
[the remark this answers is not in the document]
Comment (ifiokjr) [reply src/Holder.tsx]: You're right. I'm so used to using `"esModuleInterop": true,` I forget it's not the default.

**Comment by danilowoz** [comment]
Really thanks @AjayPoshak for your contribution! 

--

You just need to resolve the conflicts, then I can merge into `develop`. Don't worry if you don't know what changes it should to keep, I'll check it before I release a new version, if there is any doubt, keep your changes. About the `yarn.lock` you can remove the content, I can remake it later. 

Again, thanks a lot 🎉 
Let's move this amazing community ⚛️

**Comment by AjayPoshak** [comment]
@danilowoz Cool, will do it.

**Comment by AjayPoshak** [comment]
I am migrating tests to typescript as well. This may take some time. Hopefully, I'll complete it by tomorrow.

File: package.json (hunk 8 above)
Comment (AjayPoshak) [reply package.json]: removed, it was needed to run mocha_test runner, but now as we have moved to jest as test runner, so we don't need it.

File: src/__tests__/uid.test.tsx [hunk 12]
```diff
@@ -1,10 +1,10 @@
-import React from "react"
-
 import uid from "../uid"
 
 describe("unique id", () => {
   const options = 100
+  // @ts-ignore To avoid adding polyfill for `fill` becoz fill is ES6 feature and our target is ES5
   const ids = new Array(options).fill(" ").map(item => uid())
+  // @ts-ignore To avoid adding polyfill for `from`
```
Comment (AjayPoshak) [inline src/__tests__/uid.test.tsx]: To use ES6 features in this file like `fill`, `from`, `Set`, we have to add polyfills. As this is just a test file, not affecting our library consumer in any way. So I added `@ts-ignore` to ignore these `ts` errors.

File: src/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (AjayPoshak) [reply src/index.ts]: done

File: src/stylized/BulletListStyle.tsx (hunk 4 above)
Comment (AjayPoshak) [reply src/stylized/BulletListStyle.tsx]: done

File: src/stylized/BulletListStyle.tsx (hunk 5 above)
Comment (AjayPoshak) [reply src/stylized/BulletListStyle.tsx]: done

File: src/Svg.tsx (hunk 9 above)
Comment (AjayPoshak) [reply src/Svg.tsx]: done

**Comment by AjayPoshak** [comment]
@danilowoz @ifiokjr I have made requested changes.  Please review.

**Pull request merged**